### PR TITLE
Feature/simd

### PR DESCRIPTION
**Pull request by antoyo**

### CODE REVIEW

File: crate_patches/0001-rand-Enable-c2-chacha-simd-feature.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+diff --git a/rand_chacha/Cargo.toml b/rand_chacha/Cargo.toml
+index 9190b7f..872cca2 100644
+--- a/rand_chacha/Cargo.toml
++++ b/rand_chacha/Cargo.toml
+@@ -24,5 +24,5 @@ ppv-lite86 = { version = "0.2.8", default-features = false }
+ 
+ [features]
+ default = ["std"]
+-std = ["ppv-lite86/std"]
++std = ["ppv-lite86/std", "ppv-lite86/simd"]
```
Comment (bjorn3) [inline crate_patches/0001-rand-Enable-c2-chacha-simd-feature.patch:19]: This patch probably isn't necessary anymore. I was able to delete it for cg_clif.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        libgccjit_version: ["libgccjit.so", "libgccjit_without_int128.so"]
+        libgccjit_version: ["libgccjit_target_builtins.so"]
```
Comment (antoyo) [inline .github/workflows/ci.yml:14]: Revert when gcc patch is merged to master.

File: Cargo.toml [hunk 3]
```diff
@@ -14,7 +14,7 @@ path = "tests/lib.rs"
 harness = false
 
 [dependencies]
-gccjit = { git = "https://github.com/antoyo/gccjit.rs" }
+gccjit = { git = "https://github.com/antoyo/gccjit.rs", branch = "feature/target-builtins" }
```
Comment (antoyo) [inline Cargo.toml:17]: Revert when this branch is merged.

File: src/base.rs [hunk 4]
```diff
@@ -78,6 +78,9 @@ pub fn compile_codegen_unit<'tcx>(tcx: TyCtxt<'tcx>, cgu_name: Symbol, supports_
         let context = Context::default();
         // TODO(antoyo): only set on x86 platforms.
         context.add_command_line_option("-masm=intel");
+        // TODO: only add the following cli argument if the feature is supported.
+        context.add_command_line_option("-mavx2");
+        //context.add_command_line_option("-mavx512f");
```
Comment (antoyo) [inline src/base.rs:83]: Test again, but it caused an illegal instruction.

File: src/common.rs [hunk 5]
```diff
@@ -392,4 +394,21 @@ impl<'gcc, 'tcx> TypeReflection<'gcc, 'tcx> for Type<'gcc> {
...
+
+            let old_type = typ;
+            typ = typ.unqualified();
+            if old_type == typ {
+                break;
+            }
+        }
+
+        false
+    }
```
Comment (antoyo) [inline src/common.rs:413]: Fix libgccjit to not require this.

File: src/lib.rs [hunk 6]
```diff
@@ -299,9 +299,17 @@ pub fn target_features(sess: &Session) -> Vec<Symbol> {
                 if sess.is_nightly_build() || gate.is_none() { Some(feature) } else { None }
             },
         )
-        .filter(|_feature| {
+        .filter(|feature| {
             // TODO(antoyo): implement a way to get enabled feature in libgccjit.
-            false
+            // Probably using the equivalent of __builtin_cpu_supports.
+            feature.contains("sse") || feature.contains("avx")
```
Comment (antoyo) [inline src/lib.rs:305]: Fix this in a future PR since it's already big.

File: src/intrinsic/simd.rs [hunk 7]
```diff
@@ -133,6 +154,216 @@ pub fn generic_simd_intrinsic<'a, 'gcc, 'tcx>(bx: &mut Builder<'a, 'gcc, 'tcx>,
...
+            in_elem == arg_tys[2],
+            "expected inserted type `{}` (element of input `{}`), found `{}`",
+            in_elem,
+            in_ty,
+            arg_tys[2]
+        );
+        let vector = args[0].immediate();
+        let index = args[1].immediate();
+        let value = args[2].immediate();
+        // FIXME: dyncast_vector() should not need unqualified().
```
Comment (antoyo) [inline src/intrinsic/simd.rs:168]: Don't forget this one.

File: src/type_.rs [hunk 8]
```diff
@@ -127,7 +138,8 @@ impl<'gcc, 'tcx> BaseTypeMethods<'tcx> for CodegenCx<'gcc, 'tcx> {
         else if typ.is_compatible_with(self.double_type) {
             TypeKind::Double
         }
-        else if typ.dyncast_vector().is_some() {
+        // FIXME: should only use dyncast_vector.
```
Comment (antoyo) [inline src/type_.rs:141]: Don't forget this one.

File: src/builder.rs [hunk 9]
```diff
@@ -1288,14 +1294,79 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
-        ];
-        let shuffle = self.context.new_function(None, FunctionType::Extern, return_type, &params, "_mm_shuffle_epi8", false);
-        self.context.new_call(None, shuffle, &[v1, v2, mask])
+        let struct_type = mask.get_type().is_struct().expect("mask of struct type");
+
+        // TODO: use __builtin_shufflevector (or its equivalent) instead of __builtin_shuffle.
+        // TODO: perhaps use __builtin_convertvector for vector casting. (This is elsewhere,
+        // though.)
+
+        // FIXME: dyncast_vector() should not need unqualified().
```
Comment (antoyo) [inline src/builder.rs:1303]: Don't forget this one.

File: src/builder.rs [hunk 10]
```diff
@@ -879,7 +883,8 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
 
     fn sext(&mut self, value: RValue<'gcc>, dest_ty: Type<'gcc>) -> RValue<'gcc> {
         // TODO(antoyo): check that it indeed sign extend the value.
-        if dest_ty.dyncast_vector().is_some() {
+        // FIXME: only use dyncast_vector.
```
Comment (antoyo) [inline src/builder.rs:886]: Don't forget this one.

File: src/base.rs (hunk 4 above)
Comment (antoyo) [reply src/base.rs:83]: It's indeed causing an illegal instruction on `vmovdqu64` in std_example.

**Pull request merged**